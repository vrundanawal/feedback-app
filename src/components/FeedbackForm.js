import React, { useState, useContext } from "react";

//import RatingSelect from "./RatingSelect";
import Button from "./shared/Button";
import Card from "./shared/Card";
import FeedbackDataService from "../services/feedback.services";
import FeedbackContext from "../context/FeedbackContext";

const FeedbackForm = () => {
  const [text, setText] = useState("");
  const [rating, setRating] = useState(null);
  const [btnDisabled, setBtnDisabled] = useState(true);
  const [message, setMessage] = useState("");
  const [selected, setSelected] = useState(null);

  const { addFeedback } = useContext(FeedbackContext);

  const handleTextChange = (e) => {
    //console.log(e.target.value);
    if (text === "") {
      setBtnDisabled(true);
      setMessage(null);
    } else if (text !== "" && text.trim().length <= 10) {
      setBtnDisabled(true);
      setMessage("Text must be at least 10 characters");
    } else {
      setMessage(null);
      setBtnDisabled(false);
    }
    setText(e.target.value);
  };

  const handleChange = (e) => {
    // console.log(typeof +e.currentTarget.value);
    setSelected(+e.currentTarget.value);
    setRating(+e.currentTarget.value);
  };

  // const handleSubmit1 = (e) => {
  //   if (e && e.preventDefault) {
  //     e.preventDefault();
  //   }
  //   // e.preventDefault();
  //   if (!rating) {
  //     window.alert("please select the rating");
  //     return;
  //   }
  //   if (text.trim().length > 10 && rating !== "") {
  //     const newFeedback = {
  //       text: text,
  //       rating: rating,
  //     };
  //     //console.log(newFeedback);
  //     addFeedback(newFeedback);
  //     setText("");
  //     setBtnDisabled(true);
  //     setRating(null);
  //     setSelected(null);
  //   }
  // };

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (!rating) {
      window.alert("please select the rating");
      return;
    }

    if (text.trim().length > 10 && rating !== "") {
      const newFeedback = {
        text: text,
        rating: rating,
      };
      console.log(newFeedback);

      try {
        //store the data to DB
        await FeedbackDataService.addFeedbackDB(newFeedback);
      } catch (error) {
        setMessage(error.message);
      }
      addFeedback(newFeedback);
      setText("");
      setBtnDisabled(true);
      setRating(null);
      setSelected(null);
    }
  };

  return (
    <Card>
      <form onSubmit={handleSubmit}>
        <h2>How would you rate your service with us?</h2>
        {/* todo - rating select component */}
        {/* <RatingSelect select={(rating) => setRating(rating)} /> */}

        <ul className="rating">
          <li>
            <input
              type="radio"
              id="num1"
              name="rating"
              value="1"
              data-testid="select-test"
              onChange={handleChange}
              checked={selected === 1}
            />
            <label htmlFor="num1">1</label>
          </li>
          <li>
            <input
              type="radio"
              id="num2"
              name="rating"
              value="2"
              onChange={handleChange}
              checked={selected === 2}
            />
            <label htmlFor="num2">2</label>
          </li>
          <li>
            <input
              type="radio"
              id="num3"
              name="rating"
              value="3"
              onChange={handleChange}
              checked={selected === 3}
            />
            <label htmlFor="num3">3</label>
          </li>
          <li>
            <input
              type="radio"
              id="num4"
              name="rating"
              value="4"
              onChange={handleChange}
              checked={selected === 4}
            />
            <label htmlFor="num4">4</label>
          </li>
          <li>
            <input
              type="radio"
              id="num5"
              name="rating"
              value="5"
              onChange={handleChange}
              checked={selected === 5}
            />
            <label htmlFor="num5">5</label>
          </li>
          <li>
            <input
              type="radio"
              id="num6"
              name="rating"
              value="6"
              onChange={handleChange}
              checked={selected === 6}
            />
            <label htmlFor="num6">6</label>
          </li>
          <li>
            <input
              type="radio"
              id="num7"
              name="rating"
              value="7"
              onChange={handleChange}
              checked={selected === 7}
            />
            <label htmlFor="num7">7</label>
          </li>
          <li>
            <input
              type="radio"
              id="num8"
              name="rating"
              value="8"
              onChange={handleChange}
              checked={selected === 8}
            />
            <label htmlFor="num8">8</label>
          </li>
          <li>
            <input
              type="radio"
              id="num9"
              name="rating"
              value="9"
              onChange={handleChange}
              checked={selected === 9}
            />
            <label htmlFor="num9">9</label>
          </li>
          <li>
            <input
              type="radio"
              id="num10"
              name="rating"
              value="10"
              onChange={handleChange}
              checked={selected === 10}
            />
            <label htmlFor="num10">10</label>
          </li>
        </ul>

        <div className="input-group">
          <input
            type="text"
            placeholder="Write a review"
            onChange={handleTextChange}
            value={text}
            data-testid="text-input"
          />
          <Button
            type="submit"
            data-testid="add-word-button"
            isDisabled={btnDisabled}
          >
            Send
          </Button>
        </div>
        {message && (
          <div data-testid="message" className="message">
            {message}
          </div>
        )}
      </form>
    </Card>
  );
};

export default FeedbackForm;
