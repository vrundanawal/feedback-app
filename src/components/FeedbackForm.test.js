import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import FeedbackForm from "./FeedbackForm";
import userEvent from "@testing-library/user-event";
import { shallow } from "enzyme";

describe("<FeedbackForm />", () => {
  it("should render FeedbackForm component", () => {
    expect(FeedbackForm).toBeTruthy();
  });

  it("render text input", () => {
    render(<FeedbackForm />);
    const inputEl = screen.getByTestId("text-input");
    expect(inputEl).toBeInTheDocument();
    expect(inputEl).toHaveAttribute("type", "text");
  });

  test("Send button should have type Submit and it should be disable by default", () => {
    const { getByRole } = render(<FeedbackForm />);
    const submitBtn = getByRole("button", { name: "Send" });
    expect(submitBtn).toHaveAttribute("type", "submit");
    expect(submitBtn).toBeDisabled();
  });

  it("should render the component onto the screen", () => {
    render(<FeedbackForm />);
    expect(screen.getByTestId("text-input")).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("Should show error message if input value is less than 10", () => {
    render(<FeedbackForm />);
    const input = screen.getByTestId("text-input");
    fireEvent.change(input, { target: { value: "Hello" } });
    expect(screen.getByRole("button")).toBeDisabled();
  });

  it("pass valid text to input field", () => {
    render(<FeedbackForm />);
    const input = screen.getByTestId("text-input");
    userEvent.type(input, "Write a Review");
    expect(screen.queryByTestId("message")).not.toBeInTheDocument();
  });

  test("if the text is less than 10 chracter or invalid", () => {
    render(<FeedbackForm />);
    const inputEl = screen.getByTestId("text-input");
    userEvent.type(inputEl, "test");
    expect(screen.getByTestId("text-input")).toHaveValue("test");
    expect(screen.queryByTestId("message")).toBeInTheDocument();
    expect(screen.queryByTestId("message").textContent).toEqual(
      "Text must be at least 10 characters"
    );
  });

  test("renders the form with selected value", () => {
    render(<FeedbackForm />);
    const selectInput = screen.getByTestId("select-test");
    fireEvent.change(selectInput, { target: { value: "1" } });
    expect(selectInput.value).toBe("1");
  });
});

//with enzyme  write a snapshot testing
test("should test feedback form component", () => {
  const wrapper = shallow(<FeedbackForm />);
  expect(wrapper).toMatchSnapshot();
});
//add feedback to list with handler
// test("should handle onSubmit handler", () => {
//   const onSubmitSpy = jest.fn();
//   const wrapper = shallow(<FeedbackForm handleAdd={onSubmitSpy} />);
//   expect(wrapper).toMatchSnapshot();
//   wrapper.find("form").simulate("submit");

//   expect(onSubmitSpy).toHaveBeenCalled();
// });
