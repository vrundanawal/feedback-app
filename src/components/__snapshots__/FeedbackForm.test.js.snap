// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test feedback form component 1`] = `
<Card
  reverse={false}
>
  <form
    onSubmit={[Function]}
  >
    <h2>
      How would you rate your service with us?
    </h2>
    <ul
      className="rating"
    >
      <li>
        <input
          checked={false}
          data-testid="select-test"
          id="num1"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="1"
        />
        <label
          htmlFor="num1"
        >
          1
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num2"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="2"
        />
        <label
          htmlFor="num2"
        >
          2
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num3"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="3"
        />
        <label
          htmlFor="num3"
        >
          3
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num4"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="4"
        />
        <label
          htmlFor="num4"
        >
          4
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num5"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="5"
        />
        <label
          htmlFor="num5"
        >
          5
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num6"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="6"
        />
        <label
          htmlFor="num6"
        >
          6
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num7"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="7"
        />
        <label
          htmlFor="num7"
        >
          7
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num8"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="8"
        />
        <label
          htmlFor="num8"
        >
          8
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num9"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="9"
        />
        <label
          htmlFor="num9"
        >
          9
        </label>
      </li>
      <li>
        <input
          checked={false}
          id="num10"
          name="rating"
          onChange={[Function]}
          type="radio"
          value="10"
        />
        <label
          htmlFor="num10"
        >
          10
        </label>
      </li>
    </ul>
    <div
      className="input-group"
    >
      <input
        data-testid="text-input"
        onChange={[Function]}
        placeholder="Write a review"
        type="text"
        value=""
      />
      <Button
        data-testid="add-word-button"
        isDisabled={true}
        type="submit"
        version="primary"
      >
        Send
      </Button>
    </div>
  </form>
</Card>
`;
